shell/src/main/java/alluxio/cli/fs/command/CpCommand.java

```diff
@@ -250,13 +250,22 @@ private void copy(AlluxioURI srcPath, AlluxioURI dstPath, boolean recursive)
    */
   private void copyFile(AlluxioURI srcPath, AlluxioURI dstPath)
       throws AlluxioException, IOException {
+    FileOutStream os = null;
     try (Closer closer = Closer.create()) {
       OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
       FileInStream is = closer.register(mFileSystem.openFile(srcPath, openFileOptions));
       CreateFileOptions createFileOptions = CreateFileOptions.defaults();
-      FileOutStream os = closer.register(mFileSystem.createFile(dstPath, createFileOptions));
+      os = closer.register(mFileSystem.createFile(dstPath, createFileOptions));
       IOUtils.copy(is, os);
       System.out.println("Copied " + srcPath + " to " + dstPath);
+    } catch (Exception e) {
```

Comment:
if we isolate an inner try-catch to just be around `IOUtils.copy`, we don't need to check whether `os` is null, or declare it outside the original try-catch.

Reply:
@aaudiber For testing the code I did the following:
```
  @Test
  public void copyAfterWorkersNotAvailable() throws Exception {
    FileSystemShellUtilsTest.resetFileHierarchy(mFileSystem);
    File testFile = new File(mLocalAlluxioCluster.getAlluxioHome() + "/testFile");
    testFile.createNewFile();
    mFsShell.run("copyFromLocal", testFile.getPath(), "/");
    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/testFile")));
    mLocalAlluxioCluster.stopWorkers();
    mFsShell.run("cp", "/testFile", "/testFile2");
    Assert.assertFalse(mFileSystem.exists(new AlluxioURI("/testFile2")));
  }
```
But for some reason it's not working, it's the right way of testing this case?

Reply:
does it fail on the final assert? To debug I suggest setting a breakpoint in `CpCommand` to see whether `cancel()` is being called.

Reply:
@aaudiber  Thanks, I found the problem. Seems that both the null check for the `FileOutStream os` and the call to ```mFileSystem.delete``` is needed:

When we attempt to run this line without workers:
```os = closer.register(mFileSystem.createFile(dstPath, createFileOptions));```

The following exception is raised:
```io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: /192.168.0.106:39087```

The os variable becomes ```null``` which causes ```os.cancel()``` to failed.

Reply:
I think they might not be necessary. As long as the stream returned by `createFile` isn't closed, no file will be created in Alluxio. So if an exception is thrown before `os` is registered with the `Closer`, no cleanup is necessary.

Reply:
Weird for some reason if I don’t manually clean up the file the test failed. I will debug it to see if I can find the reason.

Reply:
@aaudiber The problem is caused by an inconsistency between the file system storage and metadata, when we call the method ```createFile```:
https://github.com/Alluxio/alluxio/blob/master/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L123

Even without workers the file is sucessfully created in the Alluxio metadata. I confirmed that by monitoring the ```underFSStorage``` directory on ```/tmp```, the file ```testFile2`` is never created in Alluxio underFSStorage but the file metadata is there:
```
mLocalAlluxioCluster.stopWorkers();
mFsShell.run("cp", "/testFile", "/testFile2");
URIStatus status = mFileSystem.getStatus(new AlluxioURI("/testFile2"));
```
Fetching for the ```URIStatus``` does not throw an exception.

Since there are no workers when we attempt to create the ```FileOutStream``` a netty exception is thrown in this line:
https://github.com/Alluxio/alluxio/blob/master/core/client/fs/src/main/java/alluxio/client/file/FileOutStream.java#L111

Now the ```os``` variable is null, so ```os.cancel()``` nevers get called, and the metadata remain in an incosistent state, if we don't manually delete the file using ```mFileSystem.delete(dstPath)```, the file will appear as it exist in Alluxio. 

@aaudiber What are your thoughts in this, should we keep the call to ````mFileSystem.delete(dstPath)``` or this need to be fixed in the ```BaseFileSystem``` class?

Reply:
I double checked and you're right: in the time between creating a file and closing it, the file appears as a 0-byte file to the rest of the system. So we really do need to call delete if we created the file but failed to find a worker.

It's unfortunate that there isn't a good way for callers to know whether `mFileSystem.create` failed before or after file creation. I think the best we can do right now is to handle `FileAlreadyExistsException` by not deleting, and handle any other sort of exception by deleting.

Reply:
@aaudiber Could it be fixed if we move the following 4 lines inside the try-catch block?
https://github.com/Alluxio/alluxio/blob/master/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L140

Reply:
Good idea. I'd like to use a separate try-catch statement to avoid confusion over which RPC the exceptions are thrown from, but I think it would be an improvement to the client to have it clean up the file if it fails while creating the `FileInStream`.

Reply:
@auudiber do you mean have a separate try-catch in the 4 lines that create the FileOutStream?

Reply:
something like
```java
try {
  return new FileOutStream(path, outStreamOptions, mFileSystemContext);
} catch (Exception e) {
  cleanup()
  throw e;
}
```

Reply:
Done